cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java

```diff
@@ -277,8 +271,34 @@ public VersionInfo getVersionInfo() throws IOException {
     }
   }
 
+  /**
+   * Returns a {@link RestHighLevelClient} for talking to elastic search.
+   */
+  private RestHighLevelClient getClient() {
+    RestHighLevelClient client = this.client;
+    if (client != null) {
+      return client;
+    }
+
+    synchronized (this) {
+      client = this.client;
+      if (client != null) {
+        return client;
+      }
```

Comment:
i am not sure why we will check null here again, since we have already checked it on line 279. We will only enter into this sync block when client is null.

Reply:
Search for properly implementing Java singleton with volatile. This check is necessary as there could be race condition when one thread is inside the synchronize block, while the other thread is waiting for the lock. When the first thread completed and the second thread enter the sync block, it have to check for null again.

Reply:
Just a follow up question on this, in the multi-threading case. Will it be more simpler to make getClient() as sync method and remove code block 278 to 280. I am guessing the reason why we do not do that way is due to performance? Correct me if i am wrong?

Reply:
Right. Because sync the method will turn every `getClient` call a sync operation.